### PR TITLE
provide functions for turning parameters into domain dimensions

### PR DESCRIPTION
**Pull request by skimo-openhub**
During the construction of a polyhedral model, it is not uncommon
for a statement instance set to be constructed from a description
in terms of the parameters at the point of the statement (including
the outer loop iterators) and a list of outer loop iterators
that will become the set dimensions of the instance set.
Prior to this PR, the only way of transforming
a description in terms of the parameters into one where
some of these parameters are reinterpreted as the set or domain dimensions,
is to call isl_*_move_dims.  However, the use of these functions
is discouraged.

This PR introduces functions for reinterpreting parameters
as set or domain dimensions.  The expected flow is that
the constraints are first formulated in terms of the parameter,
then an isl_multi_id object (a new type introduced in this PR)
is created that specifies which parameters should no longer
be considered parameters but should instead become part
of the instance set description.  This isl_multi_id object can
then be used to lift isl_set and isl_aff objects to refer
to the instance set.

The naming of the functions is meant to suggest that
the parameters are set free, i.e., that they are no longer
tied to a fixed (but unknown) value.
Suggestions for better names are welcome.

A potential application of these functions is available
in TC's `test/unbind` branch.

### CODE REVIEW

**Comment by tobig** [comment]
Sven, this looks very interesting. I did not have time to look in detail, but would like to go through this. I especially want to understand why you need a multi_id and a id_list is not sufficient. Hope to get to this later tonight.

**Comment by ftynse** [comment]
5da2fe996686acadbb27388bffcbae9d77f7f900: typo "These functions"

**Comment by ftynse** [comment]
430ec6a67b0653c4acf45ec52dc543c127794582: says "Conceptually, this declaration belongs to ISL_DECLARE_MULTI_DROP_DIMS," as if ISL_DECLARE_MULTI_DROP_DIMS already existed.  But the commit introduces this macro.  Not sure what was meant by this.

**Comment by ftynse** [comment]
430ec6a67b0653c4acf45ec52dc543c127794582: why do we need dummy functions on space-less multis?
are these functions exposed?
why do we return the object without erroring when dropping dims from a multi without domain?  if we had tried to drop dimensions from a zero-dimensional domain, it would have complained presumably

**Comment by ftynse** [comment]
430ec6a67b0653c4acf45ec52dc543c127794582: "make isl_multi_*_intersect_explicit_domain_product as (possibly) unused"  did you mean "mark"?

File: doc/user.pod [hunk 1]
```diff
@@ -2520,6 +2533,9 @@ Besides sets and relation, C<isl> also supports various types of functions.
 Each of these types is derived from the value type (see L</"Values">)
 or from one of two primitive function types
 through the application of zero or more type constructors.
+As a special case, a multiple expression can also be derived
+from an identifier (see L</"Identifiers">) although the result
+is not really a function.
```
Comment (ftynse) [inline doc/user.pod]: Neither is a `multi_val` (as opposed to a `multi_aff` with constant values)

File: include/isl/id_type.h [hunk 2]
```diff
@@ -12,6 +12,9 @@ typedef struct isl_id isl_id;
 
 ISL_DECLARE_EXPORTED_LIST_TYPE(id)
 
+struct __isl_export isl_multi_id;
```
Comment (ftynse) [inline include/isl/id_type.h]: `isl_multi_id` is introduced directly as exported
previous PRs seemed to export everything in separate commits to simplify upstreaming

File: isl_space.c [hunk 3]
```diff
@@ -1937,6 +1937,74 @@ __isl_give isl_space *isl_space_named_set_from_params_id(
 	return space;
 }
 
+/* Insert a domain tuple in "space" corresponding to the set space "domain".
+ * In particular, if "space" is a parameter space, then the result
+ * is the set space "domain".  If "space" is a set space, then the result
```
Comment (ftynse) [inline isl_space.c]: but the parameters are kept from "space"?  so the resulting space is not really the same as "domain"?

File: isl_space.c [hunk 4]
```diff
@@ -1937,6 +1937,74 @@ __isl_give isl_space *isl_space_named_set_from_params_id(
...
+		isl_space_free(space);
+		space = domain;
+	} else {
+		space = isl_space_map_from_domain_and_range(domain, space);
+	}
+	return space;
+}
+
+/* Internal function that introduces a domain in "space"
+ * corresponding to the tuple space of "tuple".
```
Comment (ftynse) [inline isl_space.c]: This seems to be the first place in the codebase where "tuple space" is used.
I don't fully understand what it refers to.  Probably worth documenting.  Is it
just the (set) space of the `multi_id`?

**Comment by skimo-openhub** [comment]
On Mon, Jun 11, 2018 at 03:43:43AM -0700, ftynse wrote:
> 430ec6a67b0653c4acf45ec52dc543c127794582: says "Conceptually, this declaration belongs to ISL_DECLARE_MULTI_DROP_DIMS," as if ISL_DECLARE_MULTI_DROP_DIMS already existed.  But the commit introduces this macro.  Not sure what was meant by this.

It's a typo.  It should be ISL_DECLARE_MULTI_DIMS.  Thanks,

skimo

**Comment by skimo-openhub** [comment]
On Mon, Jun 11, 2018 at 03:44:31AM -0700, ftynse wrote:
> 430ec6a67b0653c4acf45ec52dc543c127794582: why do we need dummy functions on space-less multis?

Do you mean e421e96fe8af6caf8d97e9ee8c5704693cb00953?
If so, then these functions are needed for the generic multi functions.

> are these functions exposed?

They are not part of the public API, if that is what you mean.

> why do we return the object without erroring when dropping dims from a multi without domain?  if we had tried to drop dimensions from a zero-dimensional domain, it would have complained presumably

The base expression doesn't keep track of a space, so nothing can go wrong
at this level.  The multi-expression takes care of any errors.

skimo

**Comment by skimo-openhub** [comment]
On Mon, Jun 11, 2018 at 03:44:36AM -0700, ftynse wrote:
> 430ec6a67b0653c4acf45ec52dc543c127794582: "make isl_multi_*_intersect_explicit_domain_product as (possibly) unused"  did you mean "mark"?

Yes.  Thanks,

skimo

**Comment by skimo-openhub** [comment]
On Mon, Jun 11, 2018 at 03:46:02AM -0700, ftynse wrote:
> ftynse commented on this pull request.
> > @@ -2520,6 +2533,9 @@ Besides sets and relation, C<isl> also supports various types of functions.
>  Each of these types is derived from the value type (see L</"Values">)
>  or from one of two primitive function types
>  through the application of zero or more type constructors.
> +As a special case, a multiple expression can also be derived
> +from an identifier (see L</"Identifiers">) although the result
> +is not really a function.
> 
> Neither is a `multi_val` (as opposed to a `multi_aff` with constant values)

A multi_val is a constant function and can be converted to a multi_aff.
A multi_id cannot be converted to a multi_aff.

> > @@ -12,6 +12,9 @@ typedef struct isl_id isl_id;
>  
>  ISL_DECLARE_EXPORTED_LIST_TYPE(id)
>  
> +struct __isl_export isl_multi_id;
> 
> `isl_multi_id` is introduced directly as exported
> previous PRs seemed to export everything in separate commits to simplify upstreaming

It's because the functions on this type get exported automatically
(perhaps that was a mistake, but here we are).
Also, the main reason that I do the exports in a separate step
is that we have no C++ bindings in mainline.  Once we have those
again, it's not unreasonable to export functions immediately as they
are introduced.

skimo

**Comment by skimo-openhub** [comment]
On Mon, Jun 11, 2018 at 10:47:17AM +0000, ftynse wrote:
> ftynse commented on this pull request.
> 
> ae7ec7f1a2243e8fe2198a86b633e8aa2a76267b: it looks like the function existed already, and its implementation was shorter.  If you prefer it this way, okay, but the commit message looks incorrect.

This is the version that is more likely to graduate to mainline.
Would it help if I remove the old version in a separate commit?

skimo

**Comment by skimo-openhub** [comment]
On Mon, Jun 11, 2018 at 03:48:02AM -0700, ftynse wrote:
> ftynse commented on this pull request.
> > @@ -1937,6 +1937,74 @@ __isl_give isl_space *isl_space_named_set_from_params_id(
>  	return space;
>  }
>  
> +/* Insert a domain tuple in "space" corresponding to the set space "domain".
> + * In particular, if "space" is a parameter space, then the result
> + * is the set space "domain".  If "space" is a set space, then the result
> 
> but the parameters are kept from "space"?  so the resulting space is not really the same as "domain"?

I'll update it to

 * In particular, if "space" is a parameter space, then the result
 * is the set space "domain" combined with the parameters of "space".            

skimo

**Comment by skimo-openhub** [comment]
On Mon, Jun 11, 2018 at 03:50:54AM -0700, ftynse wrote:
> ftynse commented on this pull request.
> > +
> +	is_params = isl_space_is_params(space);
> +	if (is_params < 0) {
> +		isl_space_free(domain);
> +		space = isl_space_free(space);
> +	} else if (is_params) {
> +		isl_space_free(space);
> +		space = domain;
> +	} else {
> +		space = isl_space_map_from_domain_and_range(domain, space);
> +	}
> +	return space;
> +}
> +
> +/* Internal function that introduces a domain in "space"
> + * corresponding to the tuple space of "tuple".
> 
> This seems to be the first place in the codebase where "tuple space" is used.
> I don't fully understand what it refers to.  Probably worth documenting.  Is it
> just the (set) space of the `multi_id`?

Yeah, although the space of the multi_id could in theory be a map space.
I'll just use "range space" then.  I'm not really happy about it
(especially in case we're dealing with a set space), but it's what I use
in the user documentation.

skimo

**Comment by skimo-openhub** [comment]
On Mon, Jun 11, 2018 at 03:51:44AM -0700, ftynse wrote:
> ftynse commented on this pull request.
> > + * is a map space with the original space as range.
> + * Parameters that correspond to the identifiers in "tuple" are removed.
> + *
> + * For pretty-printing purposes, the identifiers of the set dimensions
> + * of the introduced domain are set to the identifiers in "tuple".
> + */
> +__isl_give isl_space *isl_space_insert_domain_from_unbound_params(
> +	__isl_take isl_space *space, __isl_keep isl_multi_id *tuple)
> +{
> +	int i, n;
> +	isl_space *tuple_space;
> +
> +	if (!space || !tuple)
> +		return isl_space_free(space);
> +	n = isl_multi_id_size(tuple);
> +	for (i = n - 1; i >= 0; --i) {
> 
> Why do we need to iterate in the inverse order here?

It's not strictly needed.  It's a (heuristic) micro-optimization.
I added this to the comment:

 *                                                                               
 * The parameters are removed in reverse order (under the assumption
 * that they appear in the same order in "multi") because
 * it is slightly more efficient to remove parameters at the end.                

skimo

**Comment by ftynse** [comment]
On 11/06/2018 13:12, skimo-openhub wrote:
> On Mon, Jun 11, 2018 at 10:47:17AM +0000, ftynse wrote:
> > ftynse commented on this pull request.
> >
> > ae7ec7f1a2243e8fe2198a86b633e8aa2a76267b: it looks like the function 
> existed already, and its implementation was shorter. If you prefer it 
> this way, okay, but the commit message looks incorrect.
>
> This is the version that is more likely to graduate to mainline.
> Would it help if I remove the old version in a separate commit?

Commit message says "add isl_id_read_from_str".  If you change that, the 
commit itself is fine as it is.

**Comment by ftynse** [comment]
On 11/06/2018 13:59, skimo-openhub wrote:
> On Mon, Jun 11, 2018 at 03:50:54AM -0700, ftynse wrote:
> >
> > This seems to be the first place in the codebase where "tuple space" 
> is used.
> > I don't fully understand what it refers to. Probably worth 
> documenting. Is it
> > just the (set) space of the `multi_id`?
>
> Yeah, although the space of the multi_id could in theory be a map space.

A map space or a wrapped map space?

> I'll just use "range space" then. I'm not really happy about it
> (especially in case we're dealing with a set space), but it's what I use
> in the user documentation.

I was mostly wondering about the reason behind the new term. Usually, 
less terms are better as long as you don't have to overload their 
meaning.  Range space would work for me here, I think.  Tuple space 
sounds confusing with tuple id in the likes of `isl_space_set_tuple_id`.

**Comment by skimo-openhub** [comment]
On Mon, Jun 11, 2018 at 01:36:30PM +0000, ftynse wrote:
> On 11/06/2018 13:59, skimo-openhub wrote:
> > On Mon, Jun 11, 2018 at 03:50:54AM -0700, ftynse wrote:
> > >
> > > This seems to be the first place in the codebase where "tuple space" 
> > is used.
> > > I don't fully understand what it refers to. Probably worth 
> > documenting. Is it
> > > just the (set) space of the `multi_id`?
> >
> > Yeah, although the space of the multi_id could in theory be a map space.
> 
> A map space or a wrapped map space?

A map space.  (For a wrapped map, it wouldn't be an issue.)

skimo

**Comment by ftynse** [comment]
This is the main review (cc @tobig)

So the `multi_id` thing is a essentially a list of ids that is _living in some space_ as opposed to an `id_list` which does not have an associated space.  The parameters of the space may or may not contain the the ids of the space.  The idea seems to work, I mostly have conceptual questions here.

I'm not fully convinced it makes sense to have in a `multi_id` that are not parameters of the space where the `multi_id` lives.   Especially when it comes to transforming a parametric set into a non-parametric one.  It creates room for tricky to catch naming errors, e.g., unbinding the non-existent parameter because of typo in its name silently creates an unconstrained dimension.  Is there a use case for this feature?

> A multi_val is a constant function and can be converted to a multi_aff.
> A multi_id cannot be converted to a multi_aff.

It almost feels like you are forcing the multi abstraction on something that is not really a multiple expression.  Or maybe the existing `multi_*` just have too many other functions that they outgrew the "list in a space" abstraction and became something more, like a vector-valued function.  Thinking that `multi_vals` are as special as `multi_id` helps, but they are not really as special because they have, e.g., arithmetic.  I understand it's code and API reuse, but it somehow feels artificial.  No formal objection though.

----

It looks like the first dozen of commits are transforming opt-out C macro mixins into opt-in, making `multi_*` only use opt-in behavior.  This looks like a goal by itself.  Does this mean that other macro-generics will go in the same direction later?

In these commits, I wonder about the chosen granularity of implementation files.  Is it necessary, for example, to separate functions `multi_*_zero` from `multi_*_nan` from `multi_arith`?  Conceptually, zeros may be a part of arithmetic and so can be NaNs.  Are there use cases where one is necessary but not the others?

----

My proposition for names is `isl_set_unbind_params` and `isl_set_unbind_params_to_domain`, with a "unbind" as a verb. It's shorter, and also `isl_set_from_*` looks like a constructor while unbinding is essentially re-interpreting an existing set.  In C++ interface, it would look as `mySet.unbind_params(myMultiId)` as opposed to `mySet.from_unbound_params(myMultiId)` which contrasts with `isl::set::from(myUnionSet)` and looks confusing as to how exactly `myMultiId` is used in construction.

Actually, one C++ interface improvement here could be an initializer-list function with syntax `mySet.unbind_params({"N", "M"})`.

---

Is there a plan to make unbinding progressive?  That is, instead of only transforming a parametric space into a set space or a set space into a map space, to _append_ unbound dimensions to the domain (whether empty or not).

Similarly, is there a plan to add a function to _bind_ some dimensions to parameters.  In this case, the former functionality can be achieved by binding all dimensions back and then unbinding them again and managed by the caller.

**Comment by skimo-openhub** [comment]
On Mon, Jun 11, 2018 at 02:27:28PM +0000, ftynse wrote:
> This is the main review (cc @tobig)
> 
> So the `multi_id` thing is a essentially a list of ids that is _living in some space_ as opposed to an `id_list` which does not have an associated space.  The parameters of the space may or may not contain the the ids of the space.  The idea seems to work, I mostly have conceptual questions here.
> 
> I'm not fully convinced it makes sense to have in a `multi_id` that are not parameters of the space where the `multi_id` lives.   Especially when it comes to transforming a parametric set into a non-parametric one.  It creates room for tricky to catch naming errors, e.g., unbinding the non-existent parameter because of typo in its name silently creates an unconstrained dimension.  Is there a use case for this feature?

Automatic parameter alignment does the same.
If you apply a binary operation on sets and one of them
has a parameter with the wrong name, then it will just be instantiated
without any constraints.
I get the feeling the automatic parameter alignment is generally considered
a good thing and there are even calls to do it on more object types.
I could make it fail if a parameter is missing, but then I think users will
just add calls to insert all the required parameters right before the unbind call.
(I know I do that on isl_aff operations, which do not perfrom
automatic parameter alignment.)
So, I think this is a case of DWIM'ing.

> > A multi_val is a constant function and can be converted to a multi_aff.
> > A multi_id cannot be converted to a multi_aff.
> 
> It almost feels like you are forcing the multi abstraction on something that is not really a multiple expression.  Or maybe the existing `multi_*` just have too many other functions that they outgrew the "list in a space" abstraction and became something more, like a vector-valued function.  Thinking that `multi_vals` are as special as `multi_id` helps, but they are not really as special because they have, e.g., arithmetic.  I understand it's code and API reuse, but it somehow feels artificial.  No formal objection though.

I agree that it's not a perfect match, but I still think it sort
of works out and it's better than any alternative I could think of.

> It looks like the first dozen of commits are transforming opt-out C macro mixins into opt-in, making `multi_*` only use opt-in behavior.  This looks like a goal by itself.  Does this mean that other macro-generics will go in the same direction later?

Yes, but only at the point where it is needed.

> In these commits, I wonder about the chosen granularity of implementation files.  Is it necessary, for example, to separate functions `multi_*_zero` from `multi_*_nan` from `multi_arith`?  Conceptually, zeros may be a part of arithmetic and so can be NaNs.  Are there use cases where one is necessary but not the others?

I don't know what other multi expressions (if any) will be added later,
so it's difficult to say.  Zero is certainly special because
it needs to construct a canonical zero object.  In particular,
isl_multi_union_pw_aff_zero only works in the zero-dimensional case
and it was arguably a mistake to introduce it in the first place.

> My proposition for names is `isl_set_unbind_params` and `isl_set_unbind_params_to_domain`, with a "unbind" as a verb. It's shorter, and also `isl_set_from_*` looks like a constructor while unbinding is essentially re-interpreting an existing set.  In C++ interface, it would look as `mySet.unbind_params(myMultiId)` as opposed to `mySet.from_unbound_params(myMultiId)` which contrasts with `isl::set::from(myUnionSet)` and looks confusing as to how exactly `myMultiId` is used in construction.

What name would you propose for isl_set_insert_domain_from_unbound_params?

skimo

**Comment by ftynse** [comment]
> Automatic parameter alignment does the same.
> ...

Ok, this makes sense in context of automatic alignment.

> What name would you propose for isl_set_insert_domain_from_unbound_params?

`isl_set_unbind_params_to_domain` or `isl_set_unbind_params_insert_domain`

**Pull request merged**